dbms/src/Functions/GatherUtils/Algorithms.h

```diff
@@ -418,15 +432,50 @@ bool sliceHasImpl(const FirstSliceType & first, const SecondSliceType & second,
                 has = true;
         }
 
-        if (has && !all)
+        if (has && search_type == ArraySearchType::Any)
             return true;
 
-        if (!has && all)
+        if (!has && search_type == ArraySearchType::All)
             return false;
-
     }
+    return search_type == ArraySearchType::All;
+}
+/// Methods to check if first array has elements from second array, overloaded for various combinations of types.
+template < typename FirstSliceType,
+           typename SecondSliceType,
+          bool (*isEqual)(const FirstSliceType &, const SecondSliceType &, size_t, size_t)>
+bool sliceHasImplSubSeq(const FirstSliceType & first, const SecondSliceType & second, const UInt8 * first_null_map, const UInt8 * second_null_map) noexcept
+{
+    const bool has_first_null_map = first_null_map != nullptr;
+    const bool has_second_null_map = second_null_map != nullptr;
 
-    return all;
+    if (second.size == 0) 
+        return true;
+
+    for (size_t i = 0; i < first.size; ++i)
+    {
+        bool has = false;
+        for (size_t j = 0; j < second.size; ++j)
+        {
+            const bool is_first_null = has_first_null_map && first_null_map[j];
+            const bool is_second_null = has_second_null_map && second_null_map[i];
+
+            if (is_first_null && is_second_null){
+                has = true;
+            }
+            else if (!is_first_null && !is_second_null && i + j < first.size && isEqual(first, second, j, i + j)) {
+                has = true;
+            }
+            else {
+                has = false;
+                break;
+            }
+        }
+        if (has){
+            return true;
+        }
+    }
+    return false;
```

Comment:
@r-zenine any possibility to do this `auto it = std::search(v2.begin(), v2.end(), v1.begin(), v1.end());`?

Reply:
@huiwang, 
I looked into it ( that was my inital strategy ) but dropped the approach for many reasons : 
1 - We would need to define a custom predicate with a convoluted logic to take into account the null masks and the fact that we should be able to compare any two types that click-house handles. My feeling is that we will loose clarity of the purpose of the code. 
2 - It would only work for numeric arrays as `NumericArraySlice` is the only one to expose a pointer to the raw data.
3 - `GenericArraySlice` does not expose a pointer or iterator to the underlying data directly. 

I tried really hard to use `stl` algorithms without success.